### PR TITLE
Completely remade the caching functionality.

### PR DESCRIPTION
**Pull request by jwestberg**
Here's a completely new, and improved, implementation of the caching functionality. Extensively unit tested and acceptance tested, and without the issues that was plaguing the previous implementation. The code is also considerably easier to maintain the previous attempt.

It is configurable in resource.properties whether or not to use the cache.

### CODE REVIEW

File: database-impl/mongodb/src/main/java/com/findwise/hydra/mongodb/MongoQuery.java [hunk 1]
```diff
@@ -172,6 +228,99 @@ public String toJson() {
...
+//		List<String> list = new ArrayList<String>();
+//		
+//		for(Map.Entry<String, Object> e : lq.getContentsEquals().entrySet()) {
+//			if(e.getValue() == Boolean.FALSE) {
+//				list.add(e.getKey());
+//			}
+//		}
+//		
+//		return list;
+//	}
```
Comment (laserval) [inline database-impl/mongodb/src/main/java/com/findwise/hydra/mongodb/MongoQuery.java]: These should be removed, right?

Comment (jwestberg) [inline database-impl/mongodb/src/main/java/com/findwise/hydra/mongodb/MongoQuery.java]: Damn it. Yes. I'll fix it!

File: database/src/main/java/com/findwise/hydra/CachingDocumentNIO.java [hunk 2]
```diff
@@ -3,74 +3,131 @@
...
+					d.setFetchedBy(t, new Date());
+				}
+			}
+			cache.add(list);
+		}
 
 		return list;
 	}
 
+	int i = 0;
```
Comment (laserval) [inline database/src/main/java/com/findwise/hydra/CachingDocumentNIO.java]: What's this?

**Comment by laserval** [comment]
This should resolve #157, correct?

**Comment by jwestberg** [comment]
Yes. It also resolves the "unfiled" issue of the old cache being terrible ;)

**Pull request merged**